File: signature-utils/ruby/signature_utils.rb
require 'openssl'
require 'json'

module SignatureUtils
  # Recursively sort hash keys for consistent JSON serialization
  # @param data [Object] The data to sort
  # @return [Object] The data with sorted keys
  def self.sort_keys(data)
    case data
    when Hash
      sorted_hash = {}
      data.keys.sort.each do |key|
        sorted_hash[key] = sort_keys(data[key])
      end
      sorted_hash
    when Array
      data.map { |item| sort_keys(item) }
    else
      data
    end
  end
  # Generate x-signature for partner integration
  # @param body [Object] The request body object
  # @param integration_id [String] Unique integration identifier
  # @param secret_key [String] Pre-shared secret key
  # @return [String] The generated signature in format: ${integrationId}.${timestamp}.${hash}
  def self.generate_signature(body, integration_id, secret_key)
    timestamp = (Time.now.to_f * 1000).to_i
    sorted_body = sort_keys(body || {})
    json_body = JSON.generate(sorted_body)
    data_to_hash = json_body + timestamp.to_s
    
    hash = OpenSSL::HMAC.hexdigest('SHA256', secret_key, data_to_hash)
    
    "#{integration_id}.#{timestamp}.#{hash}"
  end
  
  # Validate x-signature for partner integration
  # @param signature [String] The signature to validate
  # @param body [Object] The request body object
  # @param integration_id [String] Unique integration identifier
  # @param secret_key [String] Pre-shared secret key
  # @param tolerance_ms [Integer] Timestamp tolerance in milliseconds (default: 300000 = 5 minutes)
  # @return [Boolean] True if signature is valid, false otherwise
  def self.validate_signature(signature, body, integration_id, secret_key, tolerance_ms = 300000)
    parts = signature.split('.')
    return false unless parts.length == 3
    
    sig_integration_id, sig_timestamp, sig_hash = parts
    
    # Verify integration ID matches
    return false unless sig_integration_id == integration_id
    
    # Verify timestamp is within tolerance
    current_time = (Time.now.to_f * 1000).to_i
    signature_time = sig_timestamp.to_i
    return false if (current_time - signature_time).abs > tolerance_ms
    
    # Generate expected hash
    sorted_body = sort_keys(body || {})
    json_body = JSON.generate(sorted_body)
    data_to_hash = json_body + sig_timestamp
    expected_hash = OpenSSL::HMAC.hexdigest('SHA256', secret_key, data_to_hash)
    
    # Compare hashes
    sig_hash == expected_hash
  rescue StandardError
    false
  end
end

File: signature-utils/csharp/SignatureUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;

namespace SignatureUtils
{
    public class SortedKeysContractResolver : DefaultContractResolver
    {
        protected override IList<JsonProperty> CreateProperties(Type type, MemberSerialization memberSerialization)
        {
            var properties = base.CreateProperties(type, memberSerialization);
            return properties.OrderBy(p => p.PropertyName).ToList();
        }
    }

    public static class SignatureGenerator
    {
        private static readonly JsonSerializerSettings sortedSettings = new JsonSerializerSettings
        {
            ContractResolver = new SortedKeysContractResolver(),
            Formatting = Formatting.None
        };
        /// <summary>
        /// Generate x-signature for partner integration
        /// </summary>
        /// <param name="body">The request body object</param>
        /// <param name="integrationId">Unique integration identifier</param>
        /// <param name="secretKey">Pre-shared secret key</param>
        /// <returns>The generated signature in format: ${integrationId}.${timestamp}.${hash}</returns>
        public static string GenerateSignature(object body, string integrationId, string secretKey)
        {
            var timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var jsonBody = JsonConvert.SerializeObject(body ?? new object(), sortedSettings);
            var dataToHash = jsonBody + timestamp;
            
            var hash = ComputeHmacSha256(dataToHash, secretKey);
            
            return $"{integrationId}.{timestamp}.{hash}";
        }
        
        /// <summary>
        /// Validate x-signature for partner integration
        /// </summary>
        /// <param name="signature">The signature to validate</param>
        /// <param name="body">The request body object</param>
        /// <param name="integrationId">Unique integration identifier</param>
        /// <param name="secretKey">Pre-shared secret key</param>
        /// <param name="toleranceMs">Timestamp tolerance in milliseconds (default: 300000 = 5 minutes)</param>
        /// <returns>True if signature is valid, false otherwise</returns>
        public static bool ValidateSignature(string signature, object body, string integrationId, 
            string secretKey, long toleranceMs = 300000)
        {
            try
            {
                var parts = signature.Split('.');
                if (parts.Length != 3)
                {
                    return false;
                }
                
                var sigIntegrationId = parts[0];
                var sigTimestamp = parts[1];
                var sigHash = parts[2];
                
                // Verify integration ID matches
                if (sigIntegrationId != integrationId)
                {
                    return false;
                }
                
                // Verify timestamp is within tolerance
                var currentTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
                if (!long.TryParse(sigTimestamp, out var signatureTime))
                {
                    return false;
                }
                
                if (Math.Abs(currentTime - signatureTime) > toleranceMs)
                {
                    return false;
                }
                
                // Generate expected hash
                var jsonBody = JsonConvert.SerializeObject(body ?? new object(), sortedSettings);
                var dataToHash = jsonBody + sigTimestamp;
                var expectedHash = ComputeHmacSha256(dataToHash, secretKey);
                
                // Compare hashes
                return sigHash == expectedHash;
            }
            catch
            {
                return false;
            }
        }
        
        private static string ComputeHmacSha256(string data, string key)
        {
            using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(key)))
            {
                var hash = hmac.ComputeHash(Encoding.UTF8.GetBytes(data));
                return BitConverter.ToString(hash).Replace("-", "").ToLowerInvariant();
            }
        }
    }
}
